File: lib/coverage.js
// SCL clear-sky coverage pass. The symmetric twin of the detection pass: instead
// of finding hot pixels, it measures — for every catalogue site under a scene's
// footprint — whether S2 got a CLOUD-FREE look at that site on that date. This is
// the honest denominator behind persistence (n_dates / n_clear_obs); detections
// alone can't give it, because the detector skips blocks with no hot pixels and so
// leaves no record of clear-but-unlit looks.
//
// One invocation = one scene. The scene's SCL band (20 m, ~5490² uint8 ≈ 30 MB) is
// read ONCE in full and the per-site windows are sampled from memory — far cheaper
// in-region than tens of thousands of per-site byte-range reads, and bounded.
//
// Per (site, scene) we emit the full SCL class histogram (classes 0–11) over a
// small window at the site, so the cloud-free rule stays a SQL-only knob: "grab
// everything from SCL, strip down in ETL." px_valid counts the non-nodata pixels
// (the denominator); a window that is entirely nodata yields no row (S2 did not
// usably see the site there).

import { GeoTIFF } from './vendor/geotiff-esm.js';
import { wgs84ToUtm, utmParams } from './geo.js';

// Half-window in pixels: 2 => a 5×5 px (~100 m) box centred on the site. The site
// is a point; this is the localisation window, matched to the ~20–50 m hotspot
// localisation error. Tune here (the SQL is unaffected — it reads the histogram).
const HALF_WIN = 2;

// SCL has 12 classes (0–11). 3 cloud-shadow, 8 cloud med-prob, 9 cloud high-prob,
// 10 thin cirrus are the cloud classes; 0 is nodata; the cloud-free rule lives in
// SQL over the stored histogram.
const N_SCL = 12;

// Read a full single-band raster into one typed array. The fast path is a single
// byte-range readRasters() over the whole image. But geotiff's BlockedSource throws
// "Cannot read properties of undefined (reading 'offset')" on a handful of SCL COGs
// when stitching the whole image (a block index missing from the fetched set) — and
// windowed reads trip the same bug, so it is the block source, not the read size.
// The fallback downloads the entire SCL.tif (≈0.2 MB — SCL is heavily deflated) and
// reads it from an in-memory buffer, bypassing BlockedSource. Only ever runs for
// those few scenes.
async function readFullBand(image, sclUrl) {
    try {
        const r = await image.readRasters();
        return r[0];
    } catch (err) {
        if (!/reading 'offset'/.test(err.message)) throw err;
        const buf = await (await fetch(sclUrl)).arrayBuffer();
        const tiff = await GeoTIFF.fromArrayBuffer(buf);
        const img = await tiff.getImage();
        const r = await img.readRasters();
        return r[0];
    }
}

/**
 * Sample the SCL band at each in-footprint site over one scene.
 *
 * @param {object} item  normalised STAC item: { date, mgrs, id|scene, epsg,
 *   sunElevation, sunAzimuth, bands: { scl } }
 * @param {number[]} bbox  [west, south, east, north] WGS84 (unused for the read —
 *   the full tile is read — kept for signature symmetry with detectImage)
 * @param {Array<{h3, lon, lat}>} sites  catalogue sites to sample
 * @returns {Promise<Array>} one row object per (site, scene) with a `hist` array
 */
export async function coverImage(item, bbox, sites) {
    const { bands, date, epsg, mgrs } = item;
    const scene = item.id ?? item.scene ?? null;
    const sunElevation = item.sunElevation ?? item.sun_elevation ?? null;
    const sunAzimuth = item.sunAzimuth ?? item.sun_azimuth ?? null;
    const sclUrl = bands?.scl;
    if (!sclUrl || !sites?.length) return [];

    const tiff = await GeoTIFF.fromUrl(sclUrl, { allowFullFile: false });
    const image = await tiff.getImage();
    const [imgMinX, imgMinY, imgMaxX, imgMaxY] = image.getBoundingBox();
    const width = image.getWidth();
    const height = image.getHeight();
    const resX = (imgMaxX - imgMinX) / width;
    const resY = (imgMaxY - imgMinY) / height;
    const { zone, isNorth } = utmParams(epsg);

    // Read the whole SCL band once; sample every site's window from memory.
    const scl = await readFullBand(image, sclUrl);

    const rows = [];
    for (const s of sites) {
        const [easting, northing] = wgs84ToUtm(s.lon, s.lat, zone, isNorth);
        const cx = Math.floor((easting - imgMinX) / resX);
        const cy = Math.floor((imgMaxY - northing) / resY);
        if (cx < 0 || cy < 0 || cx >= width || cy >= height) continue; // outside this tile's raster

        const hist = new Array(N_SCL).fill(0);
        let valid = 0; // non-nodata pixels = the denominator
        for (let dy = -HALF_WIN; dy <= HALF_WIN; dy++) {
            const y = cy + dy;
            if (y < 0 || y >= height) continue;
            const row = y * width;
            for (let dx = -HALF_WIN; dx <= HALF_WIN; dx++) {
                const x = cx + dx;
                if (x < 0 || x >= width) continue;
                const v = scl[row + x];
                if (v >= 0 && v < N_SCL) {
                    hist[v]++;
                    if (v !== 0) valid++; // class 0 = nodata, excluded from px_valid
                }
            }
        }
        if (valid === 0) continue; // no usable SCL at the site on this date

        rows.push({
            h3: s.h3, date, mgrs, scene,
            sun_elevation: sunElevation, sun_azimuth: sunAzimuth,
            px_valid: valid, hist,
        });
    }
    return rows;
}

File: core/src/coverage.rs
//! scl clear-sky coverage sampling — the pure compute half of lib/coverage.js.
//! given a full scl band (read by the i/o layer) it samples a small window at
//! every in-footprint catalogue site and emits the class histogram (the honest
//! n_clear_obs denominator behind persistence). the cloud-free rule stays a sql
//! knob downstream — we store the whole histogram.

use crate::geo::{utm_params, wgs84_to_utm};

// half-window in pixels: 2 ⇒ a 5×5 px (~100 m) box centred on the site.
const HALF_WIN: i64 = 2;
// scl has 12 classes (0–11); 0 is nodata.
pub const N_SCL: usize = 12;

// the clear-sky rule (permian sql/30): a look is clear when cloud_frac ≤ this.
pub const CLEAR_MAX: f64 = 0.10;
// the fixed ~100 m geographic grid the cloud mask snaps onto (3 dp ≈ the 5×5 @ 20 m
// window). detection samples cells on this grid; clustering snaps each anchor onto
// the SAME grid → the join hits. the ONLY hard rule is both sides snap identically.
pub const GRID_STEP: f64 = 0.001;
pub fn snap(x: f64) -> f64 {
    (x / GRID_STEP).round() * GRID_STEP
}
/// the cloud-mask join key: lon/lat snapped to the grid, formatted at the grid's 3 dp
/// (keep in step with GRID_STEP). a grid value re-snapped is idempotent, so detection
/// (grid centres) and clustering (snapped anchors) format to the same string.
pub fn cell_key(lon: f64, lat: f64) -> String {
    format!("{:.3},{:.3}", snap(lon), snap(lat))
}

/// grid of cell-centre sites tiling `bbox` [w,s,e,n] at GRID_STEP, each h3 its
/// `cell_key` — the cloud-mask sampling sites for one scan window.
pub fn grid_sites(bbox: [f64; 4]) -> Vec<Site> {
    let [w, s, e, n] = bbox;
    let k = |x: f64| (snap(x) / GRID_STEP).round() as i64;
    let mut sites = Vec::new();
    for j in k(s)..=k(n) {
        let lat = j as f64 * GRID_STEP;
        for i in k(w)..=k(e) {
            let lon = i as f64 * GRID_STEP;
            sites.push(Site {
                h3: cell_key(lon, lat),
                lon,
                lat,
            });
        }
    }
    sites
}

#[derive(Clone, Debug)]
pub struct Site {
    pub h3: String,
    pub lon: f64,
    pub lat: f64,
}

#[derive(Clone, Debug)]
pub struct CoverRow {
    pub h3: String,
    pub px_valid: u32,
    pub hist: [u32; N_SCL],
}

impl CoverRow {
    /// cloud fraction = (shadow + cloud med + cloud high + cirrus) / valid px — the
    /// single clear-sky classifier shared by the detection cloud mask and the cluster
    /// join, so the two can't drift. `cloud_frac ≤ CLEAR_MAX` ⇒ a clear look.
    pub fn cloud_frac(&self) -> f64 {
        if self.px_valid == 0 {
            return 0.0;
        }
        (self.hist[3] + self.hist[8] + self.hist[9] + self.hist[10]) as f64 / self.px_valid as f64
    }
}

/// sample the scl band at each in-footprint site. `img_bbox` = [min_x, min_y,
/// max_x, max_y] (utm); width/height in pixels; epsg the tile's utm code.
pub fn cover_sites(
    scl: &[u8],
    width: usize,
    height: usize,
    img_bbox: [f64; 4],
    epsg: i32,
    sites: &[Site],
) -> Vec<CoverRow> {
    let [min_x, min_y, max_x, max_y] = img_bbox;
    let res_x = (max_x - min_x) / width as f64;
    let res_y = (max_y - min_y) / height as f64;
    let (zone, is_north) = utm_params(epsg);
    let (w, h) = (width as i64, height as i64);

    let mut rows = Vec::new();
    for s in sites {
        let (easting, northing) = wgs84_to_utm(s.lon, s.lat, zone, is_north);
        let cx = ((easting - min_x) / res_x).floor() as i64;
        let cy = ((max_y - northing) / res_y).floor() as i64;
        if cx < 0 || cy < 0 || cx >= w || cy >= h {
            continue;
        }

        let mut hist = [0u32; N_SCL];
        let mut valid = 0u32;
        for dy in -HALF_WIN..=HALF_WIN {
            let y = cy + dy;
            if y < 0 || y >= h {
                continue;
            }
            let row = (y * w) as usize;
            for dx in -HALF_WIN..=HALF_WIN {
                let x = cx + dx;
                if x < 0 || x >= w {
                    continue;
                }
                let v = scl[row + x as usize] as usize;
                if v < N_SCL {
                    hist[v] += 1;
                    if v != 0 {
                        valid += 1;
                    }
                }
            }
        }
        if valid == 0 {
            continue;
        }
        rows.push(CoverRow {
            h3: s.h3.clone(),
            px_valid: valid,
            hist,
        });
    }
    rows
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::geo::{utm_params, wgs84_to_utm};

    // a site at a tile's centre samples its 5×5 (HALF_WIN=2) window; a site off the
    // tile is dropped (in-footprint filter); class 0 (nodata) doesn't count as valid.
    #[test]
    fn samples_site_window() {
        let (lon, lat, epsg) = (-102.0, 32.0, 32613); // utm 13N
        let (zone, is_north) = utm_params(epsg);
        let (e, n) = wgs84_to_utm(lon, lat, zone, is_north);
        let (w, h) = (20usize, 20usize);
        let bbox = [e - 100.0, n - 100.0, e + 100.0, n + 100.0]; // 10 m px, centre at (10,10)
        let mut scl = vec![6u8; w * h]; // all water (valid, not cloud)
        scl[0] = 0; // a nodata pixel outside the central window — must not be counted
        let sites = vec![
            Site {
                h3: "centre".into(),
                lon,
                lat,
            },
            Site {
                h3: "offtile".into(),
                lon: lon + 1.0,
                lat,
            },
        ];
        let rows = cover_sites(&scl, w, h, bbox, epsg, &sites);
        assert_eq!(rows.len(), 1, "off-tile site dropped");
        assert_eq!(rows[0].h3, "centre");
        assert_eq!(rows[0].px_valid, 25); // full 5×5, none nodata
        assert_eq!(rows[0].hist[6], 25);
        assert_eq!(rows[0].hist[0], 0);
    }

    // the grid tiles the bbox at GRID_STEP, cell keys are snapped+stable, and an
    // anchor inside a cell snaps back to that cell's key (the join contract).
    #[test]
    fn grid_and_key() {
        let sites = grid_sites([51.000, 25.000, 51.002, 25.001]); // 3 lon × 2 lat
        assert_eq!(sites.len(), 6);
        assert_eq!(sites[0].h3, "51.000,25.000");
        // an anchor 0.3 cell off a grid centre snaps to that centre's key.
        assert_eq!(cell_key(51.0013, 25.0004), "51.001,25.000");
        // the classifier: 4 cloud/shadow of 25 valid → 0.16 (not clear).
        let mut hist = [0u32; N_SCL];
        hist[6] = 21;
        hist[3] = 1;
        hist[8] = 1;
        hist[9] = 1;
        hist[10] = 1;
        let r = CoverRow {
            h3: "x".into(),
            px_valid: 25,
            hist,
        };
        assert!((r.cloud_frac() - 0.16).abs() < 1e-9);
        assert!(r.cloud_frac() > CLEAR_MAX);
    }
}
